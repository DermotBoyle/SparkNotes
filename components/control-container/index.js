export { ControlContainer } from './control-container.jsx'

export { SideMenu } from './side-menu.jsx'

export { ControlMenu } from './control-menu.jsx'
