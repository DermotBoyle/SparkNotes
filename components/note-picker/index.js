export { NotePicker } from './note-picker'
