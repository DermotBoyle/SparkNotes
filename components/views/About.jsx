import React from 'react'
import { useEffect, useState } from 'react'

export const About = ({ transitionStyles, defaultStyle, transitionState }) => {
  const [on, setOn] = useState('entering')

  useEffect(() => {
    setOn(transitionState)
  }, [])

  return (
    <section
      style={{
        ...defaultStyle,
        ...transitionStyles[on]
      }}
    >
      <p>This is an about App I built with NextJS and MongoDB</p>
    </section>
  )
}
