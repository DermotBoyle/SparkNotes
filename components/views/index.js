export { About } from './About'

export { Find } from './Find'

export { ViewAll } from './ViewAll'

export { Home } from './Home'

export { CreateNoteForm } from './CreateNoteForm'

export { EditNoteForm } from './EditNoteForm'
