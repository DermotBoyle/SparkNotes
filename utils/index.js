export * from './CONSTANTS'

export * from './helpers'

export { transitionManager } from './transition-manager/transition-manager'
