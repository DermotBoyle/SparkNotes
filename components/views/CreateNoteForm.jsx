import { SaveButton } from 'components/save-button'
import { useReducer } from 'react'
import router from 'next/router'

import styles from 'styles/create-note-form.module.scss'

import moment from 'moment'

import { Routes } from 'utils'

export const CreateNoteForm = ({
  transitionStyles,
  defaultStyle,
  transitionState
}) => {
  const initialFormValues = {
    subject: '',
    keywords: '',
    content: '',
    date: ''
  }

  const [formValues, setFormValues] = useReducer(
    (curVals, newVals) => ({ ...curVals, ...newVals }),
    initialFormValues
  )

  const { subject, keywords, content } = formValues

  const handleFormChange = ({ target }) => {
    const { name, value } = target
    setFormValues({ [name]: value })
  }

  const handleSaveClick = async () => {
    formValues.date = moment(new Date()).format('YYYY-MM-DD-T00:00:00.000Z')
    updateDB(formValues)
  }

  const updateDB = async formValues => {
    try {
      await fetch('api/note', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(formValues)
      }).then(({ statusText }) => {
        if (statusText === 'Created') {
          setFormValues(initialFormValues)
          if (res.status === 201) router.push(Routes.BASE + Routes.VIEW_ALL)
        }
      })
    } catch (error) {
      console.error(error)
    }
  }

  return (
    <section
      className={styles['form-container']}
      style={{
        ...defaultStyle,
        ...transitionStyles[transitionState]
      }}
    >
      <form>
        <div className={styles['keywords-container']}>
          <div>
            <label>Subject</label>
            <input
              type='text'
              name='subject'
              value={subject}
              onChange={handleFormChange}
            />
          </div>
          <div>
            <label>Keywords</label>
            <input
              type='text'
              name='keywords'
              value={keywords}
              onChange={handleFormChange}
            />
          </div>
        </div>
        <div className={styles['main-note']}>
          <label>Write your note here</label>
          <textarea
            type='text'
            name='content'
            value={content}
            onChange={handleFormChange}
          />
        </div>
      </form>
      <SaveButton formValues={formValues} handleSaveClick={handleSaveClick} />
    </section>
  )
}
