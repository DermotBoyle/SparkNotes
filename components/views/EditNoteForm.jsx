import { SaveButton } from 'components/save-button'
import { useReducer } from 'react'

import styles from 'styles/create-note-form.module.scss'

import { useAppContext } from 'context/global-state'
import Router from 'next/router'
import { Routes, Methods } from 'utils'

import moment from 'moment'
import { v4 as uuidv4 } from 'uuid'

export const EditNoteForm = () => {
  const {
    state: {
      subject: initialSubject,
      keywords: initialKeywords,
      content: initialContent,
      _id
    },
    dispatch
  } = useAppContext()

  const [formValues, setFormValues] = useReducer(
    (curVals, newVals) => ({ ...curVals, ...newVals }),
    {
      content: initialContent,
      keywords: initialKeywords,
      subject: initialSubject,
      currentKeyword,
      _id
    }
  )

  const { subject, keywords, content, currentKeyword } = formValues

  const handleFormChange = ({ target }) => {
    const { name, value } = target
    setFormValues({ [name]: value })
  }

  const formatKeywords = ({ keyCode, target: { value } }) => {
    if (keyCode === 32 || keyCode === 13) {
      setFormValues({ keywords: [...keywords, value], currentKeyword: '' })
    }
  }

  const handleSaveClick = async () => {
    formValues.date = moment.utc().format()
    updateDB(formValues)
  }

  const handleRemoveKeyword = value => {
    const updatedKeywords = formValues.keywords.filter(word => word !== value)
    setFormValues({ keywords: updatedKeywords })
  }

  /* const updateDB = async formValues => {
    try {
      await fetch(`api/note/${formValues._id}`, {
        method: Methods.PUT,
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(formValues)
      }).then(res => {
        if (res.status === 201) Router.push(Routes.BASE + Routes.VIEW_ALL)
      })
    } catch (error) {
      alert(error)
    }
  }*/

  const updateDB = ( formValues ) => {
    dispatch({
      type: 'updateAtId',
      payload: {currentNote: formValues}
    })
    Router.push(Routes.BASE + Routes.VIEW_ALL)
  }

  return (
    <section
      className={styles['form-container']}
    >
      <form>
        <div className={styles['inputs-container']}>
          <div className={styles['subject-input']}>
            <label>Subject</label>
            <input
              type='text'
              name='subject'
              value={subject}
              onChange={handleFormChange}
            />
          </div>
          <div className={styles['keywords-input']}>
            <label>Keywords</label>
            <p className={styles['keywords-saved']}>{keywords?.length}/ 5</p>
            <input
              disabled={keywords?.length === 5}
              type='text'
              name='currentKeyword'
              value={currentKeyword}
              onChange={handleFormChange}
              onKeyDown={formatKeywords}
            />
            <div className={styles['keyword-chips-container']}>
              {!!formValues?.keywords?.length
                ? keywords?.map(word => (
                    <p onClick={() => handleRemoveKeyword(word)} key={uuidv4()}>
                      {word} X
                    </p>
                  ))
                : null}
            </div>
          </div>
        </div>
        <div className={styles['main-note']}>
          <label>Write your note here</label>
          <textarea
            type='text'
            name='content'
            value={content}
            onChange={handleFormChange}
          />
        </div>
      </form>
      <SaveButton formValues={formValues} handleSaveClick={handleSaveClick} />
    </section>
  )
}
