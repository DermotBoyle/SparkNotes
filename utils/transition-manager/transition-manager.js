import React from 'react'
import {
  Home,
  ViewAll,
  About,
  Find,
  CreateNoteForm,
  EditNoteForm
} from 'components/views'
import { Routes } from 'utils'

export const transitionManager = () => {
  const View = {
    [Routes.HOME]: (
      <Home/>
    ),
    [Routes.VIEW_ALL]: (
      <ViewAll />
    ),
    [Routes.FIND]: (
      <Find/>
    ),
    [Routes.ABOUT]: (
      <About/>
    ),
    [Routes.CREATE]: (
      <CreateNoteForm/>
    ),
    [Routes.EDIT]: (
      <EditNoteForm/>
    )
  }

  return function ComponentToRender ({ currentURL }) {
    return View[currentURL]
  }
}

