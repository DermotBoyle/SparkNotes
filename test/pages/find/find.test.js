import React from 'react'
import { Find } from 'components/views'
import { transitionStyles } from 'utils/transition-manager/transition-manager'
import { fireEvent, render, act, screen } from '@testing-library/react'
import { WrapperProviderComp, allNotes } from "../../test.utils"

describe('<Find/>', () => {
 

    it('should find two notes in the store', () => {
        const mockDispatch = jest.fn()
        const mockState = { currentURL: '/find', savedNotes: allNotes }
        const { baseElement } = render(
          WrapperProviderComp(
            mockDispatch,
            mockState
          )(<Find transitionStyles={transitionStyles} />)
        )
    
        const input = screen.getByRole("searchbox")
        const button = screen.getByRole("button")
       

        fireEvent.change(input, { target: {value:"keywords" }})
        fireEvent.click(button)

        const listItem = screen.getAllByRole("listitem")

        expect(listItem).toHaveLength(2)
        expect(baseElement).toMatchSnapshot()
      })

      it('should find one note in the store', () => {
        const mockDispatch = jest.fn()
        const mockState = { currentURL: '/find', savedNotes: allNotes }
        const { baseElement } = render(
          WrapperProviderComp(
            mockDispatch,
            mockState
          )(<Find transitionStyles={transitionStyles} />)
        )
    
        const input = screen.getByRole("searchbox")
        const button = screen.getByRole("button")
       

        fireEvent.change(input, { target: {value:"mock" }})
        fireEvent.click(button)

        const listItem = screen.getAllByRole("listitem")

        expect(listItem).toHaveLength(1)
      })

      it('should NOT find any note in the store', () => {
        const mockDispatch = jest.fn()
        const mockState = { currentURL: '/find', savedNotes: allNotes }
        const { baseElement } = render(
          WrapperProviderComp(
            mockDispatch,
            mockState
          )(<Find transitionStyles={transitionStyles} />)
        )
    
        const input = screen.getByRole("searchbox")
        const button = screen.getByRole("button")
       

        fireEvent.change(input, { target: {value:"sdfxbdsf" }})
        fireEvent.click(button)

        const listItem = screen.queryByTestId("list-item")

        expect(listItem).toBeNull()
      })
})