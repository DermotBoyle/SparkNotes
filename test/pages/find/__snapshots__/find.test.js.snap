// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Find/> should find two notes in the store 1`] = `
<body>
  <div>
    <section
      class="find"
    >
      <label>
        Find by keyword
      </label>
      <span
        class="input-container"
        data-testid="find-component"
      >
        <input
          placeholder="Search for a keyword..."
          type="search"
        />
        <button
          name=""
          role="button"
        >
          <div
            style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="box-sizing: border-box; display: block; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                role="presentation"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTgiIGhlaWdodD0iMTgiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
              />
            </div>
            <noscript />
            <img
              alt="search icon"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
          </div>
        </button>
      </span>
      <ul>
        <li
          data-testid="list-item"
        >
          <p>
            Note: testSubject - Tuesday, June 13th, 2017 2:41:20 AM
          </p>
        </li>
        <li
          data-testid="list-item"
        >
          <p>
            Note: testSubject - Tuesday, June 13th, 2017 2:41:20 AM
          </p>
        </li>
      </ul>
    </section>
  </div>
</body>
`;
